src/crimson/os/seastore/cache.cc

```diff
@@ -1523,15 +1523,20 @@ void Cache::complete_commit(
 	   i->prior_instance);
     i->on_delta_write(final_block_start);
     i->pending_for_transaction = TRANS_ID_NULL;
-    i->prior_instance = CachedExtentRef();
     i->state = CachedExtent::extent_state_t::DIRTY;
     assert(i->version > 0);
-    if (i->version == 1 || i->get_type() == extent_types_t::ROOT) {
+    if (i->version == 1 || i->get_type() == extent_types_t::ROOT ||
+       // if the extent has been overwritten most recently, need to reset dirty_from_or_retired_at
+       (i->prior_instance &&
+       i->prior_instance->dirty_from_or_retired_at == JOURNAL_SEQ_MIN &&
+       i->state == CachedExtent::extent_state_t::CLEAN &&
+       i->version > 1)) {
```

Comment:
I'm not following why the above combination of condition can define that `i` has been overritten most recently?

Due to line 1526 above which has always set the state to DIRTY, which unconditionally bypasses this condition, so it seems to me that all the tests can pass without this commit. Does it mean that this commit is unnecessary?

Reply:
Sorry, this 1532 should be `i->prior_instance->state == CachedExtent::extent_state_t::CLEAN`. This is probably a potential issue (error happens sometimes) because `dirty ` is managed in a `dirty_from_or_retired_at ` order---this PR call `remove_from_dirty` after in-place rewrite occurs, so that proper `dirty_from_or_retired_at `needs to be set here.

Reply:
If https://github.com/ceph/ceph/pull/54525#discussion_r1423318916 makes sense, I suppose no specializations are needed here as well as replaying.

Reply:
Removed.